ヘイズ除去処理(RemoveHaze)のCUDA実装

RemoveHazeで行っているヘイズ除去処理をCUDAで実装する。入出力画像はGpuMatで管理する。

環境光を配列に格納し、CUDAカーネル内で環境光を参照する場合は、当然GPUデバイスにデータを転送する必要がある。

https://github.com/ken1714/dehazing-sample/blob/74b36d042f148c146224b415a0dc7b809137e152/dehazing-sample/src/remove_haze.cu#L38-L42

これを行わない場合、下記のようにメモリアクセスに関するエラーが発生してしまう(downloadのタイミングでエラーとなっていたため、GPUデバイスへのデータ転送が原因であることに気づくのが遅れた)。

```
4: Test command: /root/projects/dehazing-sample/dehazing-sample/build/tests/test_remove_haze
4: Test timeout computed to be: 10000000
4: Running main() from gmock_main.cc
4: [==========] Running 2 tests from 2 test suites.
4: [----------] Global test environment set-up.
4: [----------] 1 test from RemoveHazeTest
4: [ RUN      ] RemoveHazeTest.RemoveHaze
4: CPU: 28.189[msec]
4: [       OK ] RemoveHazeTest.RemoveHaze (62 ms)
4: [----------] 1 test from RemoveHazeTest (62 ms total)
4: 
4: [----------] 1 test from RemoveHazeTestCUDA
4: [ RUN      ] RemoveHazeTestCUDA.RemoveHazeCUDA
4: GPU: 0.412[msec]
4: unknown file: Failure
4: C++ exception with description "OpenCV(4.5.4) /installs/opencv-4.5.4/modules/core/src/cuda/gpu_mat.cu:249: error: (-217:Gpu API call) an illegal memory access was encountered in function 'download'
4: " thrown in the test body.
4: [  FAILED  ] RemoveHazeTestCUDA.RemoveHazeCUDA (144 ms)
4: [----------] 1 test from RemoveHazeTestCUDA (144 ms total)
4: 
4: [----------] Global test environment tear-down
4: [==========] 2 tests from 2 test suites ran. (207 ms total)
4: [  PASSED  ] 1 test.
4: [  FAILED  ] 1 test, listed below:
4: [  FAILED  ] RemoveHazeTestCUDA.RemoveHazeCUDA
4: 
4:  1 FAILED TEST
```

--- a/dehazing-sample/CMakeLists.txt
+++ b/dehazing-sample/CMakeLists.txt
@@ -2,6 +2,7 @@ cmake_minimum_required(VERSION 3.1)
 project(dehazing-sample)
 
 set(CMAKE_CXX_STANDARD 17)
+set(CUDA_NVCC_FLAGS "${CUDA_NVCC_FLAGS}; -std=c++17")
 
 # Dependencies
 find_package(OpenCV REQUIRED)
@@ -17,7 +18,7 @@ include_directories(include)
 add_subdirectory(src)
 
 # Executable file
-add_executable(main_dehazing main_dehazing.cpp)
+cuda_add_executable(main_dehazing main_dehazing.cpp)
 target_link_libraries(main_dehazing ${OpenCV_LIBRARIES} ${Boost_LIBRARIES} dehazing)
 
 # Test code

--- a/dehazing-sample/include/remove_haze.cuh
+++ b/dehazing-sample/include/remove_haze.cuh
@@ -4,8 +4,8 @@
 
 
 __global__ void dehazeGPU(unsigned char* const inputValue, unsigned char* const transmission, unsigned char* dehazedValue,
-                          const double minTransmission, const double atmosphericB, const double atmosphericG, const double atmosphericR,
+                          const double minTransmission, double* const atmosphericLight,
                           const int imageWidth, const int imageHeight);
 void executeDehazeGPU(unsigned char* inputValue, unsigned char* const transmission, unsigned char* dehazedValue,
-                      const double minTransmission, const double atmosphericB, const double atmosphericG, const double atmosphericR,
+                      const double minTransmission, double* const atmosphericLight,
                       const int imageWidth, const int imageHeight);

--- a/dehazing-sample/include/remove_haze.cuh
+++ b/dehazing-sample/include/remove_haze.cuh
@@ -1,0 +1,11 @@
+#pragma once
+
+#include "cuda_runtime.h"
+
+
+__global__ void dehazeGPU(unsigned char* const inputValue, unsigned char* const transmission, unsigned char* dehazedValue,
+                          const double minTransmission, const double atmosphericB, const double atmosphericG, const double atmosphericR,
+                          const int imageWidth, const int imageHeight);
+void executeDehazeGPU(unsigned char* inputValue, unsigned char* const transmission, unsigned char* dehazedValue,
+                      const double minTransmission, const double atmosphericB, const double atmosphericG, const double atmosphericR,
+                      const int imageWidth, const int imageHeight);

--- a/dehazing-sample/include/remove_haze.cuh
+++ b/dehazing-sample/include/remove_haze.cuh
@@ -4,8 +4,8 @@
 
 
 __global__ void dehazeGPU(unsigned char* const inputValue, unsigned char* const transmission, unsigned char* dehazedValue,
-                          const double minTransmission, double* const atmosphericLight,
+                          const float minTransmission, float* const atmosphericLight,
                           const int imageWidth, const int imageHeight);
 void executeDehazeGPU(unsigned char* inputValue, unsigned char* const transmission, unsigned char* dehazedValue,
-                      const double minTransmission, double* const atmosphericLight,
+                      const float minTransmission, float* const atmosphericLight,
                       const int imageWidth, const int imageHeight);

--- a/dehazing-sample/include/remove_haze.hpp
+++ b/dehazing-sample/include/remove_haze.hpp
@@ -6,6 +6,7 @@ class RemoveHaze: public ImageProcess {
     public:
         RemoveHaze(const double minTransmission);
         void execute(const cv::Mat& inputImage, const cv::Mat& transmissionMap, const cv::Vec3b atmosphericLight, cv::Mat& outputImage);
+        void execute(const cv::cuda::GpuMat& inputImage, const cv::cuda::GpuMat& transmissionMap, const cv::Vec3b atmosphericLight, cv::cuda::GpuMat& outputImage);
 
     protected:
         double _minTransmission;

--- a/dehazing-sample/src/CMakeLists.txt
+++ b/dehazing-sample/src/CMakeLists.txt
@@ -1,3 +1,4 @@
-file(GLOB SRCS *.cpp)
+file(GLOB SRC_FILES *.cpp)
+file(GLOB SRC_KERNELS *.cu)
 
-add_library(dehazing SHARED ${SRCS})
+cuda_add_library(dehazing ${SRC_FILES} ${SRC_KERNELS} SHARED)

--- a/dehazing-sample/src/remove_haze.cpp
+++ b/dehazing-sample/src/remove_haze.cpp
@@ -41,6 +41,9 @@ void RemoveHaze::execute(const cv::Mat& inputImage, const cv::Mat& transmissionM
 void RemoveHaze::execute(const cv::cuda::GpuMat& inputImage, const cv::cuda::GpuMat& transmissionMap, const cv::Vec3b atmosphericLight, cv::cuda::GpuMat& outputImage) {
     const int imageHeight = inputImage.rows;
     const int imageWidth  = inputImage.cols;
+    double atmosphericLightArray[3] = {(double)(unsigned char)atmosphericLight[0],
+                                       (double)(unsigned char)atmosphericLight[1],
+                                       (double)(unsigned char)atmosphericLight[2]};
 
-    executeDehazeGPU(inputImage.data, transmissionMap.data, outputImage.data, _minTransmission, atmosphericLight[0], atmosphericLight[1], atmosphericLight[2], imageWidth, imageHeight);
+    executeDehazeGPU(inputImage.data, transmissionMap.data, outputImage.data, _minTransmission, atmosphericLightArray, imageWidth, imageHeight);
 }

--- a/dehazing-sample/src/remove_haze.cpp
+++ b/dehazing-sample/src/remove_haze.cpp
@@ -41,9 +41,9 @@ void RemoveHaze::execute(const cv::Mat& inputImage, const cv::Mat& transmissionM
 void RemoveHaze::execute(const cv::cuda::GpuMat& inputImage, const cv::cuda::GpuMat& transmissionMap, const cv::Vec3b atmosphericLight, cv::cuda::GpuMat& outputImage) {
     const int imageHeight = inputImage.rows;
     const int imageWidth  = inputImage.cols;
-    double atmosphericLightArray[3] = {(double)(unsigned char)atmosphericLight[0],
-                                       (double)(unsigned char)atmosphericLight[1],
-                                       (double)(unsigned char)atmosphericLight[2]};
+    float atmosphericLightArray[3] = {(float)(unsigned char)atmosphericLight[0],
+                                      (float)(unsigned char)atmosphericLight[1],
+                                      (float)(unsigned char)atmosphericLight[2]};
 
     executeDehazeGPU(inputImage.data, transmissionMap.data, outputImage.data, _minTransmission, atmosphericLightArray, imageWidth, imageHeight);
 }

--- a/dehazing-sample/src/remove_haze.cpp
+++ b/dehazing-sample/src/remove_haze.cpp
@@ -1,4 +1,5 @@
 #include "remove_haze.hpp"
+#include "remove_haze.cuh"
 
 
 RemoveHaze::RemoveHaze(const double minTransmission): _minTransmission(minTransmission) {
@@ -34,4 +35,12 @@ void RemoveHaze::execute(const cv::Mat& inputImage, const cv::Mat& transmissionM
             }
         }
     }
+}
+
+
+void RemoveHaze::execute(const cv::cuda::GpuMat& inputImage, const cv::cuda::GpuMat& transmissionMap, const cv::Vec3b atmosphericLight, cv::cuda::GpuMat& outputImage) {
+    const int imageHeight = inputImage.rows;
+    const int imageWidth  = inputImage.cols;
+
+    executeDehazeGPU(inputImage.data, transmissionMap.data, outputImage.data, _minTransmission, atmosphericLight[0], atmosphericLight[1], atmosphericLight[2], imageWidth, imageHeight);
 }

--- a/dehazing-sample/src/remove_haze.cu
+++ b/dehazing-sample/src/remove_haze.cu
@@ -2,52 +2,44 @@
 
 
 __global__ void dehazeGPU(unsigned char* const inputValue, unsigned char* const transmission, unsigned char* dehazedValue,
-                          const double minTransmission, const double atmosphericB, const double atmosphericG, const double atmosphericR,
+                          const double minTransmission, double* const atmosphericLight,
                           const int imageWidth, const int imageHeight) {
 
     int channelNum = 3;
     int xIdx = blockIdx.x * blockDim.x + threadIdx.x;
     int yIdx = blockIdx.y * blockDim.y + threadIdx.y;
+    int zIdx = threadIdx.z;
 
-    if (xIdx >= imageWidth || yIdx >= imageHeight) return;
+    if (xIdx >= imageWidth || yIdx >= imageHeight || zIdx >= channelNum) return;
 
     double transmissionValue = transmission[yIdx * imageWidth + xIdx] / 255.0;
     transmissionValue = (transmissionValue > minTransmission) ? transmissionValue: minTransmission;
 
-    double dehazedB = ((double)(unsigned char)inputValue[(yIdx * imageWidth + xIdx) * channelNum + 0] - atmosphericB) / transmissionValue + atmosphericB;
-    double dehazedG = ((double)(unsigned char)inputValue[(yIdx * imageWidth + xIdx) * channelNum + 1] - atmosphericG) / transmissionValue + atmosphericG;
-    double dehazedR = ((double)(unsigned char)inputValue[(yIdx * imageWidth + xIdx) * channelNum + 2] - atmosphericR) / transmissionValue + atmosphericR;
+    double dehazed = ((double)(unsigned char)inputValue[(yIdx * imageWidth + xIdx) * channelNum + zIdx] - atmosphericLight[zIdx]) / transmissionValue + atmosphericLight[zIdx];
 
-    if (dehazedB > 255) {
-        dehazedB = 255;
-    } else if (dehazedB < 0) {
-        dehazedB = 0;
+    if (dehazed > 255) {
+        dehazed = 255;
+    } else if (dehazed < 0) {
+        dehazed = 0;
     }
 
-    if (dehazedG > 255) {
-        dehazedG = 255;
-    } else if (dehazedG < 0) {
-        dehazedG = 0;
-    }
-
-    if (dehazedR > 255) {
-        dehazedR = 255;
-    } else if (dehazedR < 0) {
-        dehazedR = 0;
-    }
-
-    dehazedValue[(yIdx * imageWidth + xIdx) * channelNum + 0] = dehazedB;
-    dehazedValue[(yIdx * imageWidth + xIdx) * channelNum + 1] = dehazedG;
-    dehazedValue[(yIdx * imageWidth + xIdx) * channelNum + 2] = dehazedR;
+    dehazedValue[(yIdx * imageWidth + xIdx) * channelNum + zIdx] = dehazed;
 }
 
 
 void executeDehazeGPU(unsigned char* const inputValue, unsigned char* const transmission, unsigned char* dehazedValue,
-                      const double minTransmission, const double atmosphericB, const double atmosphericG, const double atmosphericR,
+                      const double minTransmission, double* const atmosphericLight,
                       const int imageWidth, const int imageHeight) {
-
-    dim3 block(64, 4);
+    // Block and grid settings
+    int channelNum = 3;
+    dim3 block(64, 4, channelNum);
     dim3 grid((imageWidth + block.x - 1) / block.x, (imageHeight + block.y - 1) / block.y);
 
-    dehazeGPU<<<grid, block>>>(inputValue, transmission, dehazedValue, minTransmission, atmosphericB, atmosphericG, atmosphericR, imageWidth, imageHeight);
+    // Copy atmospheric light from host to device
+    double *dAtmosphericLight;
+    int atmosphericLightMemSize = sizeof(double) * channelNum;
+    cudaMalloc((void **)&dAtmosphericLight, atmosphericLightMemSize);
+    cudaMemcpy(dAtmosphericLight, atmosphericLight, atmosphericLightMemSize, cudaMemcpyHostToDevice);
+
+    dehazeGPU<<<grid, block>>>(inputValue, transmission, dehazedValue, minTransmission, dAtmosphericLight, imageWidth, imageHeight);
 }

--- a/dehazing-sample/src/remove_haze.cu
+++ b/dehazing-sample/src/remove_haze.cu
@@ -1,5 +1,4 @@
 #include "remove_haze.cuh"
-#include <stdio.h>
 
 
 __global__ void dehazeGPU(unsigned char* const inputValue, unsigned char* const transmission, unsigned char* dehazedValue,

--- a/dehazing-sample/src/remove_haze.cu
+++ b/dehazing-sample/src/remove_haze.cu
@@ -2,7 +2,7 @@
 
 
 __global__ void dehazeGPU(unsigned char* const inputValue, unsigned char* const transmission, unsigned char* dehazedValue,
-                          const double minTransmission, double* const atmosphericLight,
+                          const float minTransmission, float* const atmosphericLight,
                           const int imageWidth, const int imageHeight) {
 
     int channelNum = 3;
@@ -12,10 +12,10 @@ __global__ void dehazeGPU(unsigned char* const inputValue, unsigned char* const 
 
     if (xIdx >= imageWidth || yIdx >= imageHeight || zIdx >= channelNum) return;
 
-    double transmissionValue = transmission[yIdx * imageWidth + xIdx] / 255.0;
+    float transmissionValue = (float)(unsigned char)transmission[yIdx * imageWidth + xIdx] / 255.0;
     transmissionValue = (transmissionValue > minTransmission) ? transmissionValue: minTransmission;
 
-    double dehazed = ((double)(unsigned char)inputValue[(yIdx * imageWidth + xIdx) * channelNum + zIdx] - atmosphericLight[zIdx]) / transmissionValue + atmosphericLight[zIdx];
+    float dehazed = ((float)(unsigned char)inputValue[(yIdx * imageWidth + xIdx) * channelNum + zIdx] - atmosphericLight[zIdx]) / transmissionValue + atmosphericLight[zIdx];
 
     if (dehazed > 255) {
         dehazed = 255;
@@ -28,7 +28,7 @@ __global__ void dehazeGPU(unsigned char* const inputValue, unsigned char* const 
 
 
 void executeDehazeGPU(unsigned char* const inputValue, unsigned char* const transmission, unsigned char* dehazedValue,
-                      const double minTransmission, double* const atmosphericLight,
+                      const float minTransmission, float* const atmosphericLight,
                       const int imageWidth, const int imageHeight) {
     // Block and grid settings
     int channelNum = 3;
@@ -36,8 +36,8 @@ void executeDehazeGPU(unsigned char* const inputValue, unsigned char* const tran
     dim3 grid((imageWidth + block.x - 1) / block.x, (imageHeight + block.y - 1) / block.y);
 
     // Copy atmospheric light from host to device
-    double *dAtmosphericLight;
-    int atmosphericLightMemSize = sizeof(double) * channelNum;
+    float *dAtmosphericLight;
+    int atmosphericLightMemSize = sizeof(float) * channelNum;
     cudaMalloc((void **)&dAtmosphericLight, atmosphericLightMemSize);
     cudaMemcpy(dAtmosphericLight, atmosphericLight, atmosphericLightMemSize, cudaMemcpyHostToDevice);
 

--- a/dehazing-sample/src/remove_haze.cu
+++ b/dehazing-sample/src/remove_haze.cu
@@ -1,0 +1,54 @@
+#include "remove_haze.cuh"
+#include <stdio.h>
+
+
+__global__ void dehazeGPU(unsigned char* const inputValue, unsigned char* const transmission, unsigned char* dehazedValue,
+                          const double minTransmission, const double atmosphericB, const double atmosphericG, const double atmosphericR,
+                          const int imageWidth, const int imageHeight) {
+
+    int channelNum = 3;
+    int xIdx = blockIdx.x * blockDim.x + threadIdx.x;
+    int yIdx = blockIdx.y * blockDim.y + threadIdx.y;
+
+    if (xIdx >= imageWidth || yIdx >= imageHeight) return;
+
+    double transmissionValue = transmission[yIdx * imageWidth + xIdx] / 255.0;
+    transmissionValue = (transmissionValue > minTransmission) ? transmissionValue: minTransmission;
+
+    double dehazedB = ((double)(unsigned char)inputValue[(yIdx * imageWidth + xIdx) * channelNum + 0] - atmosphericB) / transmissionValue + atmosphericB;
+    double dehazedG = ((double)(unsigned char)inputValue[(yIdx * imageWidth + xIdx) * channelNum + 1] - atmosphericG) / transmissionValue + atmosphericG;
+    double dehazedR = ((double)(unsigned char)inputValue[(yIdx * imageWidth + xIdx) * channelNum + 2] - atmosphericR) / transmissionValue + atmosphericR;
+
+    if (dehazedB > 255) {
+        dehazedB = 255;
+    } else if (dehazedB < 0) {
+        dehazedB = 0;
+    }
+
+    if (dehazedG > 255) {
+        dehazedG = 255;
+    } else if (dehazedG < 0) {
+        dehazedG = 0;
+    }
+
+    if (dehazedR > 255) {
+        dehazedR = 255;
+    } else if (dehazedR < 0) {
+        dehazedR = 0;
+    }
+
+    dehazedValue[(yIdx * imageWidth + xIdx) * channelNum + 0] = dehazedB;
+    dehazedValue[(yIdx * imageWidth + xIdx) * channelNum + 1] = dehazedG;
+    dehazedValue[(yIdx * imageWidth + xIdx) * channelNum + 2] = dehazedR;
+}
+
+
+void executeDehazeGPU(unsigned char* const inputValue, unsigned char* const transmission, unsigned char* dehazedValue,
+                      const double minTransmission, const double atmosphericB, const double atmosphericG, const double atmosphericR,
+                      const int imageWidth, const int imageHeight) {
+
+    dim3 block(64, 4);
+    dim3 grid((imageWidth + block.x - 1) / block.x, (imageHeight + block.y - 1) / block.y);
+
+    dehazeGPU<<<grid, block>>>(inputValue, transmission, dehazedValue, minTransmission, atmosphericB, atmosphericG, atmosphericR, imageWidth, imageHeight);
+}

--- a/dehazing-sample/tests/CMakeLists.txt
+++ b/dehazing-sample/tests/CMakeLists.txt
@@ -4,7 +4,7 @@ add_subdirectory(googletest)
 # Build macro
 function(build_test test_name test_src)
     # Executable file
-    add_executable(${test_name} ${test_src} src/utils.cpp)
+    cuda_add_executable(${test_name} ${test_src} src/utils.cpp)
     target_link_libraries(${test_name} ${OpenCV_LIBRARIES} ${Boost_LIBRARIES} dehazing gmock gtest gmock_main)
     add_test(NAME ${test_name} COMMAND ${test_name} WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
 endfunction(build_test)

--- a/dehazing-sample/tests/test_remove_haze.cpp
+++ b/dehazing-sample/tests/test_remove_haze.cpp
@@ -1,14 +1,14 @@
 #include "gtest/gtest.h"
 #include "remove_haze.hpp"
+#include "remove_haze.cuh"
 #include "utils.hpp"
 
+// Path
+const std::string inputImagePath = "data/original.png";
+const std::string transmissionPath = "data/transmission_map.png";
+const std::string referenceImagePath = "data/dehazed.png";
+
 TEST(RemoveHazeTest, RemoveHaze) {
-    // Path
-    const std::string inputImagePath = "data/original.png";
-    const std::string transmissionPath = "data/transmission_map.png";
-    const std::string referenceImagePath = "data/dehazed.png";
-
-
     // Read images
     const cv::Mat inputImage = cv::imread(inputImagePath);
     const cv::Mat transmissionMap = cv::imread(transmissionPath, cv::IMREAD_GRAYSCALE);
@@ -21,5 +21,35 @@ TEST(RemoveHazeTest, RemoveHaze) {
     std::unique_ptr<RemoveHaze> removeHaze(new RemoveHaze(minTransmission));
     removeHaze->execute(inputImage, transmissionMap, atmosphericLight, outputImage);
 
+    EXPECT_EQ(equal2Images(outputImage, referenceImage), true);
+}
+
+TEST(RemoveHazeTestCUDA, RemoveHazeCUDA) {
+    // Read images
+    const cv::Mat inputImage = cv::imread(inputImagePath);
+    const cv::Mat transmissionMap = cv::imread(transmissionPath, cv::IMREAD_GRAYSCALE);
+    const cv::Mat referenceImage = cv::imread(referenceImagePath);
+    cv::Mat outputImage;
+
+    cv::cuda::GpuMat dInputImage(inputImage);
+    cv::cuda::GpuMat dTransmissionMap(transmissionMap);
+
+    const size_t stepSize = dInputImage.step * sizeof(unsigned char);
+    const size_t memSize  = stepSize * dInputImage.rows;
+
+    float *outputImagePtr;
+    cudaMalloc((void**)&outputImagePtr, memSize);
+    cv::cuda::GpuMat dOutputImage(inputImage.rows, inputImage.cols, CV_8UC3, (void*)outputImagePtr, stepSize);
+
+    dInputImage.upload(inputImage);
+    dTransmissionMap.upload(transmissionMap);
+
+    // Execute
+    const double minTransmission = 0.1;
+    const cv::Vec3b atmosphericLight = cv::Vec3b(255, 252, 249);
+    std::unique_ptr<RemoveHaze> removeHaze(new RemoveHaze(minTransmission));
+    removeHaze->execute(dInputImage, dTransmissionMap, atmosphericLight, dOutputImage);
+
+    dOutputImage.download(outputImage);
     EXPECT_EQ(equal2Images(outputImage, referenceImage), true);
 }

